pkg/operations/kubernetesupgrade/upgrader.go

```diff
@@ -216,51 +215,86 @@ func (ku *Upgrader) upgradeAgentPools() error {
+		upgradeVMs := make(map[int]*UpgradeVM)
 
 		for _, vm := range *agentPool.UpgradedAgentVMs {
 			ku.logger.Infof("Agent VM: %s, pool name: %s on expected orchestrator version\n", *vm.Name, *agentPool.Name)
 			agentIndex, _ := armhelpers.GetVMNameIndex(vm.StorageProfile.OsDisk.OsType, *vm.Name)
-			upgradedAgentsIndex[agentIndex] = true
+			upgradeVMs[agentIndex] = &UpgradeVM{*vm.Name, true}
 		}
 
 		ku.logger.Infof("Starting upgrade of agent nodes in pool identifier: %s, name: %s...\n",
 			*agentPool.Identifier, *agentPool.Name)
 
 		for _, vm := range *agentPool.AgentVMs {
-			ku.logger.Infof("Upgrading Agent VM: %s, pool name: %s\n", *vm.Name, *agentPool.Name)
-
 			agentIndex, _ := armhelpers.GetVMNameIndex(vm.StorageProfile.OsDisk.OsType, *vm.Name)
+			upgradeVMs[agentIndex] = &UpgradeVM{*vm.Name, false}
+		}
 
-			err := upgradeAgentNode.DeleteNode(vm.Name)
-			if err != nil {
-				ku.logger.Infof("Error deleting agent VM: %s\n", *vm.Name)
-				return err
-			}
+		// Create an auxiliary node if hasn't been created yet.
+		// This node is used to take on the load from deleting nodes
+		addedNode := false
+		// In a normal mode of operation the actual number of VMs in the pool [len(upgradeVMs)] is equal to agentCount.
+		// However, if the upgrade failed in the middle, the actual number of VMs might be agentCount+1 to include auxiliary node.
+		// In the later case we don't create an extra node.
+		if agentCount > 0 && len(upgradeVMs) == agentCount {
+			addedNode = true
+			agentIndex := getAvailableIndex(upgradeVMs)
+			ku.logger.Infof("Adding auxiliary agent node with index %d", agentIndex)
 
 			err = upgradeAgentNode.CreateNode(*agentPool.Name, agentIndex)
 			if err != nil {
-				ku.logger.Infof("Error creating upgraded agent VM: %s\n", *vm.Name)
+				ku.logger.Infof("Error creating auxiliary node\n")
 				return err
 			}
+			// It is possible to reconstruct VM name, but we don't have to.
+			// The VM name is need for nodes that require upgrade. However this node already has desired orchestrator version.
+			upgradeVMs[agentIndex] = &UpgradeVM{"", true}
+		}
+
+		// Upgrade nodes in agent pool
+		upgradedCount, toBeUpgraded := 0, len(*agentPool.AgentVMs)
+		for agentIndex, vm := range upgradeVMs {
+			if vm.Upgraded {
+				continue
+			}
+			ku.logger.Infof("Upgrading Agent VM: %s, pool name: %s\n", vm.Name, *agentPool.Name)
 
-			err = upgradeAgentNode.Validate(vm.Name)
+			err := upgradeAgentNode.DeleteNode(&vm.Name)
 			if err != nil {
-				ku.logger.Infof("Error validating upgraded agent VM: %s, err: %v\n", *vm.Name, err)
+				ku.logger.Infof("Error deleting agent VM: %s\n", vm.Name)
 				return err
 			}
 
-			upgradedAgentsIndex[agentIndex] = true
+			// do not create last node in favor of auxiliary node
+			if addedNode && upgradedCount == toBeUpgraded-1 {
+				ku.logger.Infof("Skipping creation of VM %s (indx %d) in favor of auxiliary node", vm.Name, agentIndex)
+				delete(upgradeVMs, agentIndex)
```

Comment:
be careful with this. Concurrent modification while looping through a map makes it very easy to skip nodes etc. Though in this case it seems like you only do it once you've determined you have already upgraded all the nodes you need to upgrade

Reply:
deleting map element while looping through it is safe.